File: euler011-proc.php
#!/usr/bin/env php
<?php
/** Problem 11: Largest product in a grid
 * https://projecteuler.net/problem=11
 */

function parseInput(string $input): array
{
    $output = [];
    $lines = explode("\n", $input);
    foreach ($lines as $line) {
        $parsedLine = [];
        $numbers = explode(" ", $line);
        foreach ($numbers as $number) {
            $parsedLine[] = intval($number);
        }
        $output[] = $parsedLine;
    }
    return $output;
}

function largestProduct(int $count, array $grid): int
{
    $size = count($grid);
    $count = 4;
    $largestProduct = 0;
    // Lines
    for ($line = 0; $line < $size; $line++) {
        for ($column = 0; $column <= $size - $count; $column++) {
            $product = 1;
            for ($offset = 0; $offset < $count; $offset++) {
                $product *= $grid[$line][$column + $offset];
            }
            if ($product > $largestProduct) {
                $largestProduct = $product;
            }
        }
    }
    // Columns
    for ($line = 0; $line <= $size - $count; $line++) {
        for ($column = 0; $column < $size; $column++) {
            $product = 1;
            for ($offset = 0; $offset < $count; $offset++) {
                $product *= $grid[$line + $offset][$column];
            }
            if ($product > $largestProduct) {
                $largestProduct = $product;
            }
        }
    }
    // Diagonals
    for ($line = 0; $line <= $size - $count; $line++) {
        for ($column = 0; $column <= $size - $count; $column++) {
            $product = 1;
            for ($offset = 0; $offset < $count; $offset++) {
                $product *= $grid[$line + $offset][$column + $offset];
            }
            if ($product > $largestProduct) {
                $largestProduct = $product;
            }
        }
    }
    // The other diagonals
    for ($line = $count - 1; $line < $size; $line++) {
        for ($column = 0; $column <= $size - $count; $column++) {
            $product = 1;
            for ($offset = 0; $offset < $count; $offset++) {
                $product *= $grid[$line - $offset][$column + $offset];
            }
            if ($product > $largestProduct) {
                $largestProduct = $product;
            }
        }
    }
    return $largestProduct;
}

function euler011(string $input): int
{
    return largestProduct(4, parseInput($input));
}

// Main
echo euler011(trim(file_get_contents('php://STDIN'))) . PHP_EOL;

File: euler011-proc.py
#!/usr/bin/env python3

# Problem 11: Largest product in a grid
# https://projecteuler.net/problem=11

import sys

def largest_product(count, grid):
    size = len(grid)
    largest_product = 0
    # Lines
    for line in range(0, size):
        for column in range(0, size - count + 1):
            product = 1
            for offset in range(0, count):
                product *= grid[line][column + offset]
            if product > largest_product:
                largest_product = product
    # Columns
    for line in range(0, size - count + 1):
        for column in range(0, size):
            product = 1
            for offset in range(0, count):
                product *= grid[line + offset][column]
            if product > largest_product:
                largest_product = product
    # Diagonals
    for line in range(0, size - count + 1):
        for column in range(0, size - count + 1):
            product = 1
            for offset in range(0, count):
                product *= grid[line + offset][column + offset]
            if product > largest_product:
                largest_product = product
    # The other diagonals
    for line in range(count - 1, size):
        for column in range(0, size - count + 1):
            product = 1
            for offset in range(0, count):
                product *= grid[line - offset][column + offset]
            if product > largest_product:
                largest_product = product
    return largest_product

def euler011(grid):
    return largest_product(4, grid)

def parse_input(lines):
    output = []
    for line in lines:
        parsed_line = []
        numbers = line.split(" ")
        for number in numbers:
            parsed_line.append(int(number))
        output.append(parsed_line)
    return output

if __name__ == "__main__":
    print(euler011(parse_input(sys.stdin.readlines())))
